src/types/ErrorRate.ts

```diff
@@ -0,0 +1,167 @@
+import { NA, RequestHealth } from './Health';
+import { RateHealthConfig, ToleranceConfig } from './ServerConfig';
+import { serverConfig } from '../config/ServerConfig';
+import { ThresholdStatus, getRequestErrorsStatus, RATIO_NA } from './Health';
+
+export interface ErrorRatio {
+  global: { status: ThresholdStatus; protocol?: string };
+  inbound: { status: ThresholdStatus; protocol?: string };
+  outbound: { status: ThresholdStatus; protocol?: string };
+}
+
+export interface Rate {
+  requestRate: number;
+  errorRate: number;
+  errorRatio: number;
+}
+
+export const regexProtocol = {
+  grpc: new RegExp(/^[1-9]$|^1[0-6]$/),
+  /* eslint-disable-next-line */
+  http: new RegExp(/^[4-5]\d\d$/)
+};
+
+const getConfig = (ns: string, name: string, kind: string): RateHealthConfig | undefined => {
+  if (serverConfig.healthConfig && serverConfig.healthConfig.rate) {
+    for (let rate of serverConfig.healthConfig.rate) {
+      if (rate.namespace.test(ns) && rate.name.test(name) && rate.kind.test(kind)) {
+        return rate;
+      }
+    }
+  }
+  return undefined;
+};
+
+export const calculateErrorRate = (
+  ns: string,
+  name: string,
+  kind: string,
+  requests: RequestHealth
+): { errorRatio: ErrorRatio; config: RateHealthConfig | undefined } => {
+  const conf = getConfig(ns, name, kind);
+  const inboundTolerances = conf?.tolerance.filter(tol => tol.direction.test('inbound'));
+  const inbound = aggregate(requests.inbound, inboundTolerances);
+  const outboundTolerances = conf?.tolerance.filter(tol => tol.direction.test('outbound'));
+  const outbound = aggregate(requests.outbound, outboundTolerances);
+  const global = calculateGlobal(inbound, outbound);
+
+  return {
+    errorRatio: {
+      global: calculateStatus(global),
+      inbound: calculateStatus(inbound, inboundTolerances),
+      outbound: calculateStatus(outbound, outboundTolerances)
+    },
+    config: conf
+  };
+};
+
+const calculateStatus = (
+  requests: { [key: string]: Rate },
+  tolerances?: ToleranceConfig[]
+): { status: ThresholdStatus; protocol: string } => {
+  let result: { status: ThresholdStatus; protocol: string } = {
+    status: {
+      value: RATIO_NA,
+      status: NA
+    },
+    protocol: ''
+  };
+  for (let [protocol, rate] of Object.entries(requests)) {
+    const auxStatus = getRequestErrorsStatus(
+      rate.errorRatio,
+      tolerances ? tolerances.filter(tol => tol.protocol.test(protocol))[0] : undefined
+    );
+    if (auxStatus.status.priority > result.status.status.priority) {
+      result.status = auxStatus;
+      result.protocol = protocol;
+    }
+  }
+  return result;
+};
+
+const aggregate = (
```

Comment:
I think this is flawed here, because this function calculates error rates and ratio depending on tolerance settings, while I think it should be independent. Tolerance is only for the status code display, it should not have any incidence on the error rate / ratio computing.

Let me take an example: you have 100 requests per second, 99 of them being success (200) and 1 being 503. The error ratio is 1/100, regardless how I configured the thresholds. Threshold configuration, however, will tell me if 1/100 errors has to be considered as failure/degraded or success health.

Reply:
PS: it would be nice to have some unit test to write the above rule in marble :)

Reply:
> Threshold configuration, however, will tell me if 1/100 errors has to be considered as failure/degraded or success health.

+1

Reply:
I calculate the error ratio in base the status code regular expression in threshold configuration.
That was the feature in this PR , be able to set what is an error for the app